### PR TITLE
[v1.x] ONNX Type inference support

### PR DESCRIPTION
**Pull request by Zha0q1**
Previously we do not have any mechanism for inferring dtypes for intermediate nodes. This is a issue for nodes such as plus_scalar, arange_like etc where we need to make explicitly set the output type based on the input type. Previously we just use a global default value but obviously this is not correct. This pr attempts to introduce proper type inference in the onnx support module

### CODE REVIEW

File: python/mxnet/contrib/onnx/mx2onnx/export_onnx.py [hunk 1]
```diff
@@ -314,7 +325,24 @@ def create_onnx_graph_proto(self, sym, params, in_shape, in_type, verbose=False,
...
                 node_output_names.sort()
-                outputs_lookup.append(node_output_names)
+                # match the output names to output dtypes
+                if dtypes is not None:
+                    assert len(node_output_names) == len(dtypes)
+                    node_outputs = [NodeOutput(node_output_names[i], dtypes[i])
+                                    for i in range(len(dtypes))]
+                else:
+                    # in case dtypes is None, we just default to the dtype of the first input
+                    #TODO
```
Comment (waytrue17) [inline python/mxnet/contrib/onnx/mx2onnx/export_onnx.py:335]: Seems no need for TODO here?

File: python/mxnet/contrib/onnx/mx2onnx/export_onnx.py [hunk 2]
```diff
@@ -314,7 +325,24 @@ def create_onnx_graph_proto(self, sym, params, in_shape, in_type, verbose=False,
...
+                    #assert len(node["inputs"]) > 0
+                    if len(node["inputs"]) > 0:
+                        first_input = node["inputs"][0]
+                        first_input_dtype = outputs_lookup[first_input[0]][first_input[1]].dtype
+                    else:
+                        #print('no inputs')
```
Comment (waytrue17) [inline python/mxnet/contrib/onnx/mx2onnx/export_onnx.py:341]: Should we remove print?

**Review by waytrue17 — APPROVED** [review]

File: python/mxnet/contrib/onnx/mx2onnx/export_onnx.py (hunk 1 above)
Comment (Zha0q1) [reply python/mxnet/contrib/onnx/mx2onnx/export_onnx.py]: Yeah I refactored that part

**Review by waytrue17 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**